From 63a0b302ee4c2b1e056bf53a627fff7b7e4dd619 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Thu, 29 Dec 2016 14:39:48 +0800
Subject: [PATCH 1448/5242] MLK-13684-4 video: mxsfb: add pan display for
 overlay fb

commit  c0610f507faf0e97050df8f85c49f26d1ec83d6c from
https://source.codeaurora.org/external/imx/linux-imx.git

Add pan display support for the AS overlay framebuffer.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxsfb.c |   34 ++++++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/drivers/video/fbdev/mxsfb.c b/drivers/video/fbdev/mxsfb.c
index e850bdb..2c35219 100644
--- a/drivers/video/fbdev/mxsfb.c
+++ b/drivers/video/fbdev/mxsfb.c
@@ -1752,6 +1752,11 @@ static int overlayfb_check_var(struct fb_var_screeninfo *var,
 		}
 		break;
 	}
+
+        if (var->xres_virtual * var->yres_virtual * var->bits_per_pixel / 8 >
+            info->fix.smem_len)
+		return -EINVAL;
+
 	fill_fmt_bitfields(var, rgb);
 
 	return 0;
@@ -1809,6 +1814,34 @@ static int overlayfb_blank(int blank, struct fb_info *info)
 	return 0;
 }
 
+static int overlayfb_pan_display(struct fb_var_screeninfo *var,
+				 struct fb_info *info)
+{
+	int ret = 0;
+	unsigned int bytes_offset;
+	struct mxsfb_layer *ofb = (struct mxsfb_layer *)info->par;
+	struct mxsfb_info  *fbi = ofb->fbi;
+
+	init_completion(&fbi->flip_complete);
+
+	bytes_offset = info->fix.line_length * var->yoffset;
+	writel(info->fix.smem_start + bytes_offset,
+	       fbi->base + LCDC_AS_NEXT_BUF);
+
+	/* update on next VSYNC */
+	writel(CTRL1_CUR_FRAME_DONE_IRQ_EN,
+	       fbi->base + LCDC_CTRL1 + REG_SET);
+
+	ret = wait_for_completion_timeout(&fbi->flip_complete, HZ / 2);
+	if (!ret) {
+		dev_err(info->device,
+			"overlay wait for pane flip timeout\n");
+		return -ETIMEDOUT;
+	}
+
+	return 0;
+}
+
 static struct fb_ops overlay_fb_ops = {
 	.owner		= THIS_MODULE,
 	.fb_open	= overlayfb_open,
@@ -1816,6 +1849,7 @@ static int overlayfb_blank(int blank, struct fb_info *info)
 	.fb_check_var	= overlayfb_check_var,
 	.fb_set_par	= overlayfb_set_par,
 	.fb_blank	= overlayfb_blank,
+	.fb_pan_display = overlayfb_pan_display,
 	.fb_mmap 	= mxsfb_mmap,
 };
 
-- 
1.7.9.5

